rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,273 @@
+
+Steps 1-3 are already possible with Teleport currently and used for various MFA
+features. Steps 4 and 5 will require some changes to the Auth API client and
+server respectively.
+
+#### Client changes
+
+There are a few different ways that the Auth client can pass MFA verification with
+the Auth server, each with their own pros and cons:
+
+1. Add the `MFAChallengeResponse` argument to each administrative API endpoint.
+
+The Auth server can consume/verify the challenge response in each request and
+treat the request as MFA verified.
+
+Pros:
+
+- It is straightforward and matches existing endpoints which require MFA
+  verification (rpc `GenerateUserSingleUseCerts`, login endpoints, etc.).
+- It is clear by the API reference which endpoints require MFA verification and
+  which do not, making it easier to reason from the client perspective, especially
+  for external users of the client.
+- The MFA verification only applies to one API request, preventing MFA-replay-like
+  attacks.
+
+Cons:
+
+- Requires custom changes to both the client and server implementations for
+  each affected endpoint, both to pass the MFA challenge response from the
+  client and for the Auth server to verify it for the request.
+  - This can potentially be avoided within the implementation by introducing
+    a common gRPC message for the `MFAChallengeResponse`.
+
+1. Allow clients to pass `MFAChallengeResponse` as client request metadata.
+
+This approach is similar to passing a bearer-token in HTTP requests. Note that it
+will augment the normal certificate auth flow, not replace it. Additionally, the
+MFA Challenge Response will be passed to the server within the context of the TLS
+handshake, so it is secure.
+
+The client can cleanly pass the challenge response with the gRPC client call
+option `PerRPCCredentials`. The Auth server can then grab the `MFAChallengeResponse`
+from the request metadata, consume/verify the challenge response, and treat the request
+as MFA verified. [POC](https://github.com/gravitational/teleport/commit/aa0a8102eccd91cff2851053fba3e1c271bdaa65).
+
+Pros:
+
+- Client can apply common logic for MFA verification, rather than endpoint
+  specific logic.
+  - Ex: First request without MFA, receive access denied error, re-request with MFA.
+  - MFA verification can be extended to additional Auth endpoint without any client changes.
+- Adds a new system of passing user credentials, which can be used in other
+  endpoints that always or sometimes need MFA verification.
+  - Ex: `GenerateUserSingleUseCerts` could be replaced by simply checking for
+    MFA verification in the request metadata of `GenerateUserCerts`.
+- The MFA verification only applies to one API request, preventing MFA-replay-like
+  attacks.
+
+Cons:
+
+- Adding a new system of passing user credentials makes the client harder to
+  reason about, especially for external users of the API client.
+
+1. Reissue short-lived (1 minute TTL) MFA verified TLS certificates.
```

Comment:
This solution could also help address the MFA per-action concern I raised above by encoding the allowed action into the cert.  Making the cert specific to a specific action can also reduce the reuse concerns.

Reply:
Actually I think it's the opposite. This solution creates the per-action concern, as listed in the cons.

You're right that per-action scoped certificates would resolve this issue, but [action-scoped certs is a separate concern](https://github.com/gravitational/teleport/issues/22140) that is not yet implemented and is out of scope for this RFD.

Reply:
I think if we choose this option we need to bring action-scoped certs into scope for this RFD.  I don't think we can reasonably implement a cert based MFA solution without encoding the action being done into the cert.

Reply:
> I don't think we can reasonably implement a cert based MFA solution without encoding the action being done into the cert.

Why? We've had cert-based MFA solutions without encoding the action into the cert for several years. Per-session MFA does exactly this, doesn't it?